src/main/java/edu/kit/ipd/crowdcontrol/objectservice/moneytransfer/MoneyTransferManager.java

```diff
@@ -0,0 +1,291 @@
+            if (!worker.getEmail().equals("")) {
+                List<GiftCodeRecord> payedCodesForWorker = chooseGiftCodes(worker, giftCodes);
+
+                payWorker(worker, payedCodesForWorker);
+                giftCodes = workerBalanceOperations.getUnusedGiftCodes();
+            }
+        }
+        LOGGER.trace("Completed submission of giftcodes to workers.");
+
+        //sends a notification about problems with submission of giftcodes
+        sendNotification(notificationText.toString(), false);
+    }
+
+    /**
+     * Fetches new giftcodes from the mailbox and saves them in the database.
+     *
+     * @throws MoneyTransferException gets thrown, if an error occurred
+     */
+    private void fetchNewGiftCodes() throws MoneyTransferException {
+        Message[] messages;
+
+        //fetch new mails
+        try {
+            messages = mailHandler.fetchUnseen("inbox");
+        } catch (MessagingException e) {
+            throw new MoneyTransferException("The MailHandler couldn't fetch new giftcodes from the mailserver. " +
+                    "It seems, that there is either a problem with the server or with the properties file.");
+        }
+
+        //extract giftcodes and save them to the database
+        for (Message message : messages) {
+            try {
+                Optional<GiftCodeRecord> rec = MailParser.parseAmazonGiftCode(message);
+                if (rec.isPresent()) {
+                    workerBalanceOperations.addGiftCode(rec.get().getCode(), rec.get().getAmount());
+                }
+            } catch (MoneyTransferException e) {
+                try {
+                    mailHandler.markAsUnseen(message);
+                    throw new MoneyTransferException(e.getMessage());
+                } catch (MessagingException f) {
+                    throw new MoneyTransferException("The MailHandler couldnt mark mails as unseen." +
+                            "It seems, that there is either a problem with the server or with the properties file." + System.getProperty("line.separator") +
+                            e.getMessage());
+                }
+            }
+        }
+    }
+
+    /**
+     * Chooses the giftcodes, a worker receives as payment.
+     *
+     * @param worker    the worker to pay
+     * @param giftCodes all giftcodes
+     * @return returns the list of the chosen giftcodes
+     */
+    private List<GiftCodeRecord> chooseGiftCodes(WorkerRecord worker, List<GiftCodeRecord> giftCodes) {
+        List<GiftCodeRecord> payedCodes = new ArrayList<>();
+        int creditBalance = workerBalanceOperations.getBalance(worker.getIdWorker());
+
+        for (GiftCodeRecord nextCode : giftCodes) {
+            if (creditBalance == 0) {
+                break;
+            }
+            if (nextCode.getAmount() <= creditBalance) {
+                payedCodes.add(nextCode);
+                creditBalance -= nextCode.getAmount();
+            }
+        }
```

Comment:
the ONLY solution to guarantee that the payment is maximal is try out all possible combinations of giftcode values. I can do this, but this will take exponential time. And this becomes to a problem, if there are many giftcodes with different values in the database.